### ISSUE TITLE
SDFGI Is Splotchy On Large Uniform Surfaces

### ISSUE THREAD

**Issue by WickedInsignia** [godotengine/godot#50773]
### Godot version

Godot 4 nightly build

### System information

Windows 10 v20H2, Intel i7 6700, Nvidia GTX1080, 16GB RAM

### Issue description

SDFGI displays severe splotching and pockmarking artifacts in low-light environments and especially on large uniform surfaces with little texture or model breakup. Although this splotching seems to always be present, it is mostly present in these circumstances.

### Steps to reproduce

In the following examples I have a room with a single thin window letting directional light in as well as skylight. This scenario has been recreated in the attached project file below.
Here, we can observe severe splotching of light and dark tones in a large bare room:
![SDFGI Splotchy 05](https://user-images.githubusercontent.com/70243134/126780033-794a4b65-523b-4b67-9f99-3adad9fa6495.JPG)
![SDFGI Splotchy 01](https://user-images.githubusercontent.com/70243134/126780227-d9ba9e5c-c943-4c8f-8326-edff8b0344e8.JPG)

Even on smaller-scale surfaces and those with low detail frequency the splotching can be observed. Although it may look smooth from a distance:
![SDFGI Splotchy 02](https://user-images.githubusercontent.com/70243134/126780307-ba73844c-bf7c-4c49-8be5-b250c09cf1a2.JPG)

Splotching occurs when the camera moves closer and GI detail is increased in the area:
![SDFGI Splotchy 03](https://user-images.githubusercontent.com/70243134/126780368-620c7bf2-c03b-4d0b-b17a-0d4699b3cad5.JPG)

### Minimal reproduction project

I have provided an example file with a couple basic meshes and a capsule to represent the size of a regular player:
[SDFGI_Splotchy_Bug.zip](https://github.com/godotengine/godot/files/6868679/SDFGI_Splotchy_Bug.zip)

Here are some screenshots of the effects that may be observed:
The leftmost structure is only partially enclosed with plenty of light spilling in. Although the effect is hardly noticeable close to the light, further into the structure where the light has difficulty reaching there are highly visible artifacts:
![SDFGI Splotchy 06](https://user-images.githubusercontent.com/70243134/126781073-4f75b6ab-c867-4b0f-a899-c2afab7502c6.JPG)

The rightmost structure is considerable larger with a thinner opening for light to enter. Although there is not detail frequency in the space, the GI lighting artifacts are chaotic:
![SDFGI Splotchy 07](https://user-images.githubusercontent.com/70243134/126781185-b48136f6-ac43-48c1-a2a9-f68e150aa20f.JPG)

All shots taken with 8 cascades, Y scale at 50% and Occlusion and Sky Influence active.

**Comment by Calinou** [comment]
You can increase SDFGI's Probe Ray Count and Frames to Converge in the Project Settings to improve this.

**Comment by WickedInsignia** [comment]
Thanks! Will check this out and record the results if they’re significant.
It might be worth making a proposal to be able to restrict SDFGI to a probe domain like GI Probe? I understand if it’s not possible since the systems are inherently different.
Also I haven’t tested lightmaps and SDFGI in unison, but if lightmapping is used in a specific space would it be possible to stop SDFGI from influencing the same area?

**Comment by Calinou** [comment]
> It might be worth making a proposal to be able to restrict SDFGI to a probe domain like GI Probe? I understand if it’s not possible since the systems are inherently different.

SDFGI is designed to be used in open world scenes first and foremost, where bounds don't really make sense and aren't really feasible on a technical level.

**Comment by WickedInsignia** [comment]
Ah good to know. Just trying to think of ways to reach the most flexible and good-looking compromise, but it seems like with some extra elbow grease SDFGI is the way to go on the whole.

**Comment by clayjohn** [comment]
> It might be worth making a proposal to be able to restrict SDFGI to a probe domain like GI Probe?

@WickedInsignia If you need more detail on an area, specifically and indoor area with a small but bright light source (like your test case), you can add a VoxelGIProbe. The VoxelGIProbe can blend over, or outright replace the GI contribution from SDFGI.

**Comment by Calinou** [comment]
If you want to play around the the number of SDFGI frames to converge associated to each setting, you can change this line and recompile the editor: https://github.com/godotengine/godot/blob/9b034c930467606a7696d93020838c4f0c5f01e2/servers/rendering/renderer_rd/renderer_scene_render_rd.cpp#L65

Note that the project settings' text will not change, but the value associated to each setting behind the scenes will be changed successfully.

**Comment by WickedInsignia** [comment]
@Calinou I am merely but an artist haha. I'll have a look into that solution but I'm using Godot as shipped since I don't have the experience to change or recompile the editor code. Optimally, the ability to get a less splotchy result would be an in-editor setting. That may be asking for a lot though!

@clayjohn I hadn't tested GIProbe with SDFGI yet so I will give that a shot. Would using both together significantly reduce performance?

**Comment by clayjohn** [comment]
@WickedInsignia It depends on your hardware and the settings you use. But you are combining the cost of both effects so it will be more expensive than using either technique by itself. Thankfully VoxelGi is not very expensive to use at runtime (it requires baking though). You can also run VoxelGI at half resolution to make it even faster at runtime.

**Comment by Calinou** [comment]
Another way to reduce splotches is to increase **Bounce Feedback** and decrease energy to compensate for the brighter result. This results in a more uniform look with more gradual fadeouts:

### Bounce Feedback 0, Energy 1

![2021-07-24_06 36 39](https://user-images.githubusercontent.com/180032/126857504-53033c40-a4b3-4d5e-a692-070a11c97d7b.png)

### Bounce Feedback 0.6, Energy 0.5

![2021-07-24_06 37 45](https://user-images.githubusercontent.com/180032/126857505-81f39d29-5fe3-4875-97cf-abd791dae045.png)

**Comment by WickedInsignia** [comment]
Nice! Seems to resolve the splotching to at least some level. Ran my own tests briefly, and it seems quite easy to hit the limits of Bounce Feedback becoming unusable, so potential users should be advised. It also took a few seconds for the adjustments to take effect so I made sure to wait until they settled.

Here's **Bounce Feedback 0.8 and Energy 0.5**. The large structure is quite nice:
![Bouncefeedback08_Energy05_01](https://user-images.githubusercontent.com/70243134/126868181-ff313f9c-ee53-49e0-994c-b22a55a6f8cc.JPG)

However the small structure is unusable and bleeds through the geometry:
![Bouncefeedback08_Energy05_02](https://user-images.githubusercontent.com/70243134/126868212-52a220f8-72fb-425d-b916-d9cf26bffdc7.JPG)

Setting it down to Bounce Feedback 0.7 and Energy 0.7 was a happy medium in this case:
![Bouncefeedback07_Energy07_01](https://user-images.githubusercontent.com/70243134/126868248-dd53643e-8da7-46c3-bf4e-4336caef38e8.JPG)
![Bouncefeedback07_Energy07_02](https://user-images.githubusercontent.com/70243134/126868249-e1d4192a-2589-4957-98b8-0d77e3c6caad.JPG)

It's also worth taking note of how drastic the SDFGI effect with this adjustments was depending on distance. The smaller structure looks vastly different at a distance:
![Bouncefeedback06_Energy05_DistanceLong](https://user-images.githubusercontent.com/70243134/126868270-bf56454c-57fe-4094-b169-351a391ae9ee.JPG)
![Bouncefeedback06_Energy05_DistanceShort](https://user-images.githubusercontent.com/70243134/126868277-5d7d58ca-e23e-4898-afca-dffbefbb296f.JPG)

I also attempted to use the VoxelGI, but found at default settings my framerate tanked and there was excess light leakage:
![AlternativeVoxelGI_01](https://user-images.githubusercontent.com/70243134/126868287-215d1b00-5b37-4e3b-9d96-ff6211de4df8.JPG)

Lastly, I attempted to use LightmapGI at default settings but encountered even worse light leakage and some streaking artifacts. The lightmap options seem to differ considerably from 3.3.2 so I didn't bother taking shots. If you get some time, try this out since you'd probably know how to operate it better than I.

In conclusion none of the 3 GI options gave a satisfyingly smooth result (in my brief tests). VoxelGI was closest, but suffered excess light leakage. The lightmapping in Godot 4 gave vastly different results to Godot 3 in my other tests, but I'm probably not using it correctly and settled on defaults.

I absolutely don't expect the tools to do "all of the work" for the artist and there will always be caveats that need to be acknowledged, which is ultimately the goal of these bug reports.

**Comment by Calinou** [comment]
> I also attempted to use the VoxelGI, but found at default settings my framerate tanked and there was excess light leakage:

- To improve performance, enable **Half Size** in the project settings. This affects both SDFGI and VoxelGI, and improves performance significantly at a small quality cost. The quality difference is less noticeable at higher resolutions. I have a pull request enabling that setting by default: https://github.com/godotengine/godot/pull/49738
- To reduce leaking, adjust the GIProbe's position/rotation/extents a bit, bake again and experiment with values that make leaking less noticeable. Since the walls are axis-aligned in this demo, I would recommend playing with just the position and extents.
  - You can also use solid meshes with bake mode set to **Static** that you hide after baking the VoxelGI.

**Comment by WickedInsignia** [comment]
Awesome!! That’s really helpful, thank you. I’ll make sure to play with all of these a lot further and these tips will help a ton in my own projects.

Also very thankful for you thoroughness, since it presents viable solutions for other artists who come across the same issues.

**Comment by DEF7** [comment]
After spending a half-hour teasing apart gi.cpp, and seeing that everything important is being done in compute shaders, I've found: [https://github.com/godotengine/godot/blob/bf9cf7d0346968ca482465ffb1fc2bf94f2ae877/servers/rendering/renderer_rd/shaders/environment/sdfgi_integrate.glsl#L187
](url)

Ideally, rays would be fired off in evenly spaced directions, both spatially (i.e. over the ray count) and temporally (across the # of frames to convergence). Unfortunately, this is not the case with the existing implementation.

Even with 1 ray fired off per probe per frame, we should have an even distribution of rays if our convergence were set to a huge number of frames, with rays all evenly distributed. Well, I went ahead and implemented Godot's current ray vector calculation to visualize the distribution of rays across the ray count and number of convergence frames, and I think what I've found will give some insight that could lead to the improvement of Godot's new SDFGI lighting system.

Here's four rays being fired across four frames in Godot's SDFGI (yes, I know four frames isn't an option in the engine, I'm working with powers of two here per personal preference):

![image](https://user-images.githubusercontent.com/11559157/205218697-a26e627b-0c10-4360-9c7b-624cbc55e2e7.png)
..

Off the bat we can see that distribution with these 16 total rays isn't very great, it's very lopsided. This is alleviated by increasing the ray count, of course, which directly impacts performance (along with the total number of probes). Also, Godot randomizes, or at least attempts to randomize, the ray vectors that each individual probe samples along - but in my experiments I've found that it simply rotates the same set of rays at a different orientation (around one axis) for each probe, rather than calculating a unique ray distribution for each. Not only that, but their orientation doesn't vary by much - they all seem to be within a very narrow angle range rather than 0-360.

Here are all of the rays Godot samples the scene with, at 1024 rays per frame, converging over four frames (i.e. 4096 rays total):

![image](https://user-images.githubusercontent.com/11559157/205219002-defed723-7da9-472b-aeae-498805511af2.png)
..

It would be nice if we could have 1024 rays per frame, per probe, but it's not a realistic number. Performance is greatly improved if we shoot fewer rays and compensate by increasing our number of frames that we converge across instead (albeit with lighting update lag as the caveat) ...but this is where things go sideways with the current implementation. Here we have 4 rays per frame, converging across 1024 frames (i.e. 4096 rays):

![image](https://user-images.githubusercontent.com/11559157/205219024-a9a9d317-baaa-47b7-aa9e-f3f0937a4b8b.png)
..

As you can see, spreading the rays out across multiple frames and averaging the result leaves us with huge swaths of the space around a probe completely missing from what would be an ideal sampling distribution. Of course, nobody is converging over 1024 frames, so lets try to find a happy medium. Here's what 32 rays per frame, converging over 32 frames looks like, from a more horizontal vantage point (i.e. 1024 rays total):

![image](https://user-images.githubusercontent.com/11559157/205219042-65dc2996-02ae-4ad1-b6d8-3aee8380544e.png)
..

When we try to balance our rays/convergence out like this there are still "strata" that are not being included in the sample distribution. Here's 32 rays per frame but converging across 1024 frames (i.e. 32768 rays):

![image](https://user-images.githubusercontent.com/11559157/205219060-3ca37430-27f8-45e8-bc20-63d0117c0759.png)
..

![image](https://user-images.githubusercontent.com/11559157/205219076-9bac9abb-2f98-4744-8bf8-b2e24b0b5434.png)
..


Godot uses the typical rendering coordinate convention where Y is vertical, which means that these visualizations have Y and Z swapped by comparison. Hence, the ray distributions you're seeing in these images are actually rotated on their side with their top/bottom hemispheres facing along the horizonal Z axis in Godot.

What this exercise demonstrates is that the ray distribution function should at least be inverted, if Godot is to continue using the vogel_hemisphere() function to calculate ray vectors for each probe each frame. **_Increasing the convergence frames that samples are averaged across should vary their vector the way that increasing the number of rays currently does, and increasing the number of rays should affect their distribution the way that increasing the convergence frames currently does._** This would allow users to have lower ray counts in their games and less splotchiness, because increasing the frame convergence would mean having more rays that are better distributed. Frame convergence would be able to do a much better job at evening the lighting out. Right now increasing convergence frames still leaves bands of unsampled strata. Flipping the way history index and ray index are fed into vogel_hemisphere() would be more ideal for Godot's SDFGI. This also might help with issues around how varied lighting appears across cascades.

Another idea is to use some kind of PRNG noise function, instead of vogel_hemisphere(), which has a uniform distribution within a variable domain (ray_count X convergence_frames ???). I'm not an expert on that but it might be able to provide a better ray distribution for light sampling than vogel_hemisphere(). While I think my proposed solution to improving ray distribution will be worthwhile, there's still a bit of clumping in the ray vectors that could be evened out with a better function.

Lastly, ray counts and/or convergence frames could also vary based on cascade, where probes in the lower cascades (closer to the camera) fire more rays per frame and converge over less frames while the higher level cascades fire less rays per frame and converge over a higher number of frames - rather than all probes firing the same number of rays per frame and converge over the same number of frames. This would improve performance by reducing the total number of rays being fired per frame. Lighting closer to the camera could be sharper and update faster while lighting on the farther scenery would not just be lower quality but also update more slowly.

Hope this helps! :)
